python/pyspark/sql/types.py

```diff
@@ -938,12 +1023,17 @@ def _infer_type(obj):
                 return MapType(_infer_type(key), _infer_type(value), True)
         else:
             return MapType(NullType(), NullType(), True)
-    elif isinstance(obj, (list, array)):
+    elif isinstance(obj, list):
         for v in obj:
             if v is not None:
                 return ArrayType(_infer_type(obj[0]), True)
         else:
             return ArrayType(NullType(), True)
+    elif isinstance(obj, array):
+        if obj.typecode in _array_type_mappings:
```

Comment:
And also keys() makes a copy of a key list in python 2 whereas in operator against a dict directly calls `__contains__` if I remember this correctly. Let's leave this out in any event.